zookeeper-specifications/zab-1.0/ZabWithFLE.tla

```diff
@@ -0,0 +1,1020 @@
+           /\ Reply(i, j, [mtype     |-> NEWLEADER,
+                           mepoch    |-> acceptedEpoch[i],
+                           mlastZxid |-> lastZxid[i]])
+        \/ /\ ~canSync
+           /\ Discard(j, i)
+           /\ UNCHANGED <<history, lastZxid, currentVote, commitIndex, synced, forwarding, ackIndex, committedCounter, proposalMsgsLog>>
+        
+(* In phase l12, leader waits for receiving ACKEPOPCH from a quorum,
+   and check whether it has the latest history and epoch from them.
+   If so, leader's zabState turns to SYNCHRONIZATION. *)
+LeaderHandleACKEPOCH(i, j) ==
+        /\ state[i] = LEADING
+        /\ msgs[j][i] /= << >>
+        /\ msgs[j][i][1].mtype = ACKEPOCH
+        /\ LET msg == msgs[j][i][1]
+               infoOk == /\ j \in learners[i]            
+                         /\ acceptedEpoch[i] = msg.mepoch
+               logOk  == /\ infoOk                          \* logOk = TRUE means leader is more up-to-date than follower j
+                         /\ \/ currentEpoch[i] > msg.mlastEpoch
+                            \/ /\ currentEpoch[i] = msg.mlastEpoch
+                               /\ \/ lastZxid[i][1] > msg.mlastZxid[1]
+                                  \/ /\ lastZxid[i][1] = msg.mlastZxid[1]
+                                     /\ lastZxid[i][2] >= msg.mlastZxid[2]
+               replyOk == /\ infoOk
+                          /\ NullPoint \in ackeRecv[i]
+           IN /\ infoOk
+              /\ \/ /\ replyOk
+                    /\ subRECOVERYSYNC(i, j)
+                    /\ ackeRecv' = [ackeRecv EXCEPT ![i] = IF j \notin ackeRecv[i] THEN ackeRecv[i] \union {j}
+                                                                                   ELSE ackeRecv[i]]
+                    /\ UNCHANGED <<state, currentEpoch,logicalClock, receiveVotes, outOfElection, recvQueue, waitNotmsg, leadingVoteSet, electionMsgs, idTable,
+                                   zabState, leaderAddr, learners>>
+                 \/ /\ ~replyOk
+                    /\ \/ /\ logOk
+                          /\ ackeRecv' = [ackeRecv EXCEPT ![i] = IF j \notin ackeRecv[i] THEN ackeRecv[i] \union {j}
+                                                                                         ELSE ackeRecv[i]]
+                          /\ Discard(j, i)
+                          /\ UNCHANGED <<varsL, zabState, leaderAddr, learners, forwarding>>
+                       \/ /\ ~logOk            \* go back to election
+                          /\ LeaderShutdown(i)
+                          /\ UNCHANGED ackeRecv
+                    /\ UNCHANGED <<history, commitIndex, synced, forwarding, ackIndex, committedCounter, proposalMsgsLog>>
+        /\ UNCHANGED <<acceptedEpoch, cepochRecv, ackldRecv,currentCounter, sendCounter, committedIndex, tempVarsZ, cepochSent, epochLeader, inherentViolated>>
+                          
+LeaderDiscovery2(i) ==
```

Comment:
Action name is not descriptive, perhaps rename to something like LeaderTransitionsToSynchronization